air-script/tests/bitwise/bitwise.masm

```diff
@@ -1,38 +1,37 @@
 proc.cache_z_exp
-    # Load trace length
-    mem_load.4294903306
-    # => [trace_len, ...]
-    # Push initial num_of_cycles
-    push.1
-    # => [num_of_cycles, trace_len, ...]
-    # Load Z
     padw mem_loadw.4294903304 drop drop
-    # => [z_1, z_0, num_of_cycles, trace_len, ...]
-    # Compute exponentiations based on the number of cycles for a period of 8
-    dup.3 div.8
-    # => [num_of_cycles', z_1, z_0, num_of_cycles, trace_len, ...]
-    # Update next num_of_cycles and compute number of iterations
-    movup.3 dup.1 movdn.4 div
-    # => [i, z_1, z_0, num_of_cycles', trace_len, ...]
-    # Exponentiate the existing `z**num_of_cycles` an additional `i` times
-    dup.0 neq.1
+    # => [z_1, z_0, ...]
+    # Number of exponentiations for a period of length 8
+    mem_load.4294903307 neg add.3
+    # => [count, z_1, z_0, ...]
+    # Exponentiate z
+    dup.0 neq.0
     while.true
         movdn.2 dup.1 dup.1 ext2mul
-        # => [z_1^2, z_0^2, i, num_of_cycles', trace_len, ...]
-        movup.2 div.2 dup.0 neq.1
-        # => [b, i+1, z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+        # => [z_1^2, z_0^2, i, ...]
+        movup.2 add.1 dup.0 neq.0
+        # => [b, i+1, z_1^2, z_0^2, ...]
     end # END while
-    drop push.0 push.0 mem_storew.500000100 # Save the exponentiation of z for column 0
-    mem_storew.500000101 # Save the exponentiation of z for column 1
-    # => [0, 0, z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+    push.0 mem_storew.500000100 # z^exp for period 8
+    # => [0, 0, z_1^2, z_0^2, ...]
     drop drop
-    # => [z_1^2, z_0^2, num_of_cycles', trace_len, ...]
+    # Exponentiate z 8 times, until trace_len
+    push.18446744069414584313 dup.0 neq.0
+    while.true
+        movdn.2 dup.1 dup.1 ext2mul
+        # => [z_1^2, z_0^2, i, ...]
+        movup.2 add.1 dup.0 neq.0
+        # => [b, i+1, z_1^2, z_0^2, ...]
+    end # END while
+    push.0 mem_storew.500000101 # z^exp for trace_len
+    # => [0, 0, z_1^2, z_0^2, ...]
+    drop drop
+    # => [z_1^2, z_0^2, num_of_cycles', -log2(trace_len), ...]
```

Comment:
I am a bit lost here, what are we computing?

Reply:
The `num_of_cycles', -log2(trace_len)` was an outdated state of the stack, sorry about that, I have updated the code to show the correct stack instead.